torch/jit/__init__.py

```diff
@@ -702,32 +682,100 @@ def init_then_register(self, *args, **kwargs):
         return super(ScriptMeta, cls).__init__(name, bases, attrs)
 
 
-class ScriptModule(with_metaclass(ScriptMeta, torch._C.ScriptModule)):
-
-    def __setattr__(self, name, value):
-        if isinstance(value, Parameter):
-            self._register_or_set_parameter(name, value)
-        elif isinstance(value, ScriptModule):
-            self._register_module(name, value)
-            # note: script modules are subclassed in python and the
-            # C++ script::Module class will not hold references to them
-            # to ensure that you always get the same python value here
-            # we store it as a native attribute _in addition to_
-            # registering it with the C++ script::Module
-            object.__setattr__(self, name, value)
-        else:
-            object.__setattr__(self, name, value)
+class ScriptModule(with_metaclass(ScriptMeta, Module, torch._C.ScriptModule)):
+    def __init__(self, optimize=True):
+        Module.__init__(self)
+        torch._C.ScriptModule.__init__(self, optimize)
+        self._parameters = OrderedParameterDict(self)
+        self._buffers = OrderedBufferDict(self)
+        self._modules = OrderedModuleDict(self)
 
     def __getattr__(self, attr):
-        r = self._get_attribute(attr)
-        if r is None:
-            raise AttributeError("'{}' object has no attribute '{}'".format(self.__class__.__name__, attr))
-        return r
+        if self._has_method(attr):
+            return self._get_method(attr)
+        return Module.__getattr__(self, attr)
+
+    # Module already has this method defined, so we
+    # need to override it and send it through the ScriptModule lookup
+    def forward(self, *args, **kwargs):
+        return self.__getattr__('forward')(*args, **kwargs)
 
     def define(self, lang):
         rcb = createResolutionCallback()
         self._define(lang, rcb, True)
 
 
+def _get_methods(cls):
+    import inspect
+    # In Python 3 unbound methods are functions, but in Python 2 they are methods
+    return inspect.getmembers(cls, predicate=lambda x: inspect.isfunction(x) or inspect.ismethod(x))
+
+
+_compiled_methods_whitelist = {
+    'forward', 'register_buffer', 'register_parameter', 'add_module',
+    '_apply', 'apply', 'cuda', 'cpu', 'type', 'float', 'double', 'half',
+    'state_dict', 'load_state_dict', 'parameters', 'named_parameters',
+    '_all_buffers', 'children', 'named_children', 'modules', 'named_modules',
+    'zero_grad', 'share_memory', '_get_name'
+}
+
+
+def _make_fail(name):
+    def fail(self, *args, **kwargs):
+        raise RuntimeError(name + " is not supported on TracedModules")
+    return fail
+
+
+for name, method in _get_methods(torch.nn.Module):
+    if name.startswith('__'):
+        continue
+    if name not in ScriptModule.__dict__ and name not in _compiled_methods_whitelist:
+        setattr(ScriptModule, method.__name__, _make_fail(name))
```

Comment:
Not sure if we want an equally strict whitelist in `ScriptModule`. It was written with `TracedModule` in mind, so it really includes only "`const` methods".

Reply:
I did go through and  edit the whitelist to set it to the things I think would work given the current code in `nn.Module`, if there are specifics functions we can add that are not in this list, we should.